library(shinythemes)
library(DT)
require(XML)
library(plyr)
library(dplyr)
library(DT)

db_set <- c("Gmax_275_Wm82.a2.v1.cds_primaryTranscriptOnly.fa",
            "Gmax_275_Wm82.a2.v1.protein_primaryTranscriptOnly.fa")
db_path <- c("~/database/Gmax_a2v1/annotation/")

ui <- fluidPage(theme = shinytheme("cerulean"),
                tagList(
                  tags$head(
                    tags$link(rel="stylesheet", type="text/css",href="style.css"),
                    tags$script(type="text/javascript", src = "busy.js")
                  )
                ),
                
                #This block gives us all the inputs:
                mainPanel(
                  headerPanel('Shiny Blast'),
                  textAreaInput('query', 'Input sequence:',
                                value = "", placeholder = "", width = "600px", height="200px"),
                  selectInput("Reference", "Select Reference", choices=c("William 82 a2v1", "W05"),
                              seleced = "William 82 a2v1",
                              width="120px"),
                  selectInput("dbtype", "Databse Type:", choices=c("Ref Type: Genome","Ref Type: Proteome"),
                              selected = "Ref Type: Genome",
                              width="120px"),
                  div(style="display:inline-block",
                      selectInput("program", "Program:", choices=c("blastn","tblastn"), width="120px")),
                  div(style="display:inline-block",
                      selectInput("eval", "e-value:", choices=c(1,0.001,1e-4,1e-5,1e-10), width="120px")),
                  actionButton("blast", "BLAST!")
                ),
                
                #this snippet generates a progress indicator for long BLASTs
                div(class = "busy",  
                    p("Calculation in progress.."), 
                    img(src="https://i.stack.imgur.com/8puiO.gif", height = 100, width = 100,align = "center")
                ),
                
                #Basic results output
                mainPanel(
                  h4("Results"),
                  DT::dataTableOutput("blastResults"),
                  p("Alignment:", tableOutput("clicked") ),
                  verbatimTextOutput("alignment")
                )
)

server <- function(input, output, session){
  
    ##custom_db <- c("LvTx")
    ##custom_db_path <- c("/LV_transcriptome/LvTx")
    ## Chunk for database selection
    observe({
        x <- input$dbtype
        if (x == "Ref Type: Genome"){
            y <- c("blastn", "tblastn")
        }else if(x== "Ref Type: Proteome"){
            y <- c("blastp")
        }else{}
    # Can also set the label and select items
        updateSelectInput(session, "program",
                          label = "Program",
                          choices = y)
      ##selected = tail(y, 1)
    })
    ## Indicate database selected
    db_selected <- reactive({
        if(input$Reference == "William 82 a2v1" && input$dbtype == "Ref Type: Genome"){
            paste0(db_path, db_set[1])
        }else if(input$Reference == "William 82 a2v1" && input$dbtype == "Ref Type: Proteome"){
            paste0(db_path, db_set[2])
        }else if(input$Reference == "W05" && input$dbtype == "Ref Type: Genome"){
            paste0(db_path, db_set[3])
        }else if(input$Reference == "William 82 a2v1" && input$dbtype == "Ref Type: Proteome"){
            paste0(db_path, db_set[4])
        }
    })
    
    blastresults <- eventReactive(input$blast, {
    
    #gather input and set up temp file
    query <- input$query
    tmp <- tempfile(fileext = ".fa")
    
    #if else chooses the right database
    ##if (input$db == custom_db){
    ##  db <- db_selected()
    ##  remote <- c("")
    ##} else {
    ##  db <- c("nr")
    ##  #add remote option for nr since we don't have a local copy
    ##  remote <- c("-remote")
    ##}
    
    #this makes sure the fasta is formatted properly
    if (startsWith(query, ">")){
      writeLines(query, tmp)
    } else {
      writeLines(paste0(">Query\n",query), tmp)
    }
    
    #calls the blast
    data <- system(paste0(input$program," -query ",tmp," -db ",db_selected()," -evalue ",input$eval," -outfmt 5 -max_hsps 1 -max_target_seqs 10 ",remote), intern = T)
    xmlParse(data)
    }, ignoreNULL= T)

  #Now to parse the results...
  parsedresults <- reactive({
    if (is.null(blastresults())){}
    else {
      xmltop = xmlRoot(blastresults())
      
      #the first chunk is for multi-fastas
      results <- xpathApply(blastresults(), '//Iteration',function(row){
        query_ID <- getNodeSet(row, 'Iteration_query-def') %>% sapply(., xmlValue)
        hit_IDs <- getNodeSet(row, 'Iteration_hits//Hit//Hit_id') %>% sapply(., xmlValue)
        hit_length <- getNodeSet(row, 'Iteration_hits//Hit//Hit_len') %>% sapply(., xmlValue)
        bitscore <- getNodeSet(row, 'Iteration_hits//Hit//Hit_hsps//Hsp//Hsp_bit-score') %>% sapply(., xmlValue)
        eval <- getNodeSet(row, 'Iteration_hits//Hit//Hit_hsps//Hsp//Hsp_evalue') %>% sapply(., xmlValue)
        cbind(query_ID,hit_IDs,hit_length,bitscore,eval)
      })
      #this ensures that NAs get added for no hits
      results <-  rbind.fill(lapply(results,function(y){as.data.frame((y),stringsAsFactors=FALSE)}))
    }
  })
  
  #makes the datatable
  output$blastResults <- renderDataTable({
    if (is.null(blastresults())){
    } else {
      parsedresults()
    }
  }, selection="single")
  
  #this chunk gets the alignemnt information from a clicked row
  output$clicked <- renderTable({
    if(is.null(input$blastResults_rows_selected)){}
    else{
      xmltop = xmlRoot(blastresults())
      clicked = input$blastResults_rows_selected
      tableout<- data.frame(parsedresults()[clicked,])
      
      tableout <- t(tableout)
      names(tableout) <- c("")
      rownames(tableout) <- c("Query ID","Hit ID", "Length", "Bit Score", "e-value")
      colnames(tableout) <- NULL
      data.frame(tableout)
    }
  },rownames =T,colnames =F)
  
  #this chunk makes the alignments for clicked rows
  output$alignment <- renderText({
    if(is.null(input$blastResults_rows_selected)){}
    else{
      xmltop = xmlRoot(blastresults())
      
      clicked = input$blastResults_rows_selected
        
      #loop over the xml to get the alignments
      align <- xpathApply(blastresults(), '//Iteration',function(row){
        top <- getNodeSet(row, 'Iteration_hits//Hit//Hit_hsps//Hsp//Hsp_qseq') %>% sapply(., xmlValue)
        mid <- getNodeSet(row, 'Iteration_hits//Hit//Hit_hsps//Hsp//Hsp_midline') %>% sapply(., xmlValue)
        bottom <- getNodeSet(row, 'Iteration_hits//Hit//Hit_hsps//Hsp//Hsp_hseq') %>% sapply(., xmlValue)
        rbind(top,mid,bottom)
      })
        
      #split the alignments every 40 carachters to get a "wrapped look"
      alignx <- do.call("cbind", align)
      splits <- strsplit(gsub("(.{40})", "\\1,", alignx[1:3,clicked]),",")
      
      #paste them together with returns '\n' on the breaks
      split_out <- lapply(1:length(splits[[1]]),function(i){
        rbind(paste0("Q-",splits[[1]][i],"\n"),paste0("M-",splits[[2]][i],"\n"),paste0("H-",splits[[3]][i],"\n"))
      })
      unlist(split_out)
    }
  })
}
#phew...
